content/posts/2023-12-optimising-spark-memory-usage.md

```diff
@@ -0,0 +1,278 @@
+---
+title: How we reduced Spark's memory usage by 90%
+date: 2023-12-14
+---
+
+## TL;DR
+
+- Convert plain data object to class instances.
+- De-duplicate immutable string values.
+- Store timestamps as numbers.
+- Choose carefully how you calculate percentiles.
+
+## Introduction
+
+After I finished [optimizing our database usage](../2023-11-optimising-spark-db-perf), another
+problem came to our attention: the spark-evaluate service was consuming too much memory. Before I
+describe our change, let me explain what this service does.
+
+SPARK operates in rounds; each round takes approximately 1 hour in the current configuration. During
+the round, SPARK checker nodes running in Stations perform retrieval checks and submit the results -
+we call these records _measurements_. After a round is over, the evaluation service processes these
+measurements to evaluate the impact of each node (each participant). The Meridian smart contract
+uses the outcome of this evaluation to schedule rewards for node operators. You can learn more about
+Meridian on GitHub: https://github.com/Meridian-IE/impact-evaluator
```

Comment:
I'm not sure this is relevant. Basically it's about processing batches of records in huge numbers. Is that enough context for this post?

Reply:
> Basically it's about processing batches of records in huge numbers. Is that enough context for this post?

That's a great way to frame this 👏🏻 

Do you feel it's distracting to explain Meridian? My goal was to spread the knowledge about Meridian and perhaps even spark some interest.

Reply:
Yes that was my feeling. With this length of post it would be hard to keep my attention, if you then dive too deep into these technologies I think you will lose lot's of readers. We can tease the technology though?